rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+*not* as [source code](glossary.html#source-code) (which is how Python does it)
+or as [binary code](glossary.html#binary-code) (which is how app stores distribute things).
+When a package is built,
+R loads and checks the code,
+then saves the corresponding low-level instructions.
+Our R files should therefore define functions,
+not run commands immediately:
+if they do the latter,
+those commands will be executed every time the script loads,
+which is probably not what users will want.
+
+As a side effect,
+this means that if a package uses `load(something)`,
+then that `load` command is executed *while the package is being compiled*,
+and *not* while the compiled package is being loaded by a user after distribution.
+Thus,
+this simple and rather pointless "package":
+
+```{r eval=FALSE}
+library(stringr)
+
+sr <- function(text, pattern, replacement) {
+  str_replace(text, pattern, replacement)
+}
+```
+
+may not work when it's loaded by a user,
+because `stringr` may not be in memory on the user's machine at the time `str_replace` is called.
+
+How then can our packages use libraries?
+The safest way is to use [fully-qualified names](glossary.html#fully-qualified-name)
+such as `stringr::str_replace`
+every time we call a function defined somewhere outside our package.
+We will explore other options below.
+
+### Exercise: What packages do you have?
+
+What R packages are currently installed on your computer?
+How did you figure this out?
+
+### Exercise: Pros and cons
+
+What are the pros and cons of distributing packages as source code, byte code, and binary code?
+
+## How do I create a package? {#rse-r-package-create}
+
+We will create an R package called `zipffreq` (with no dashes or other special characters in its name)
+to hold word frequencies in classic English novels
+and some functions to merge and plot that data.
+The first step is to run RStudio's project creation wizard,
+select "New Project",
+and then tell RStudio to make it a package (and a Git repository):
+
+FIXME: figure
+
+Before doing our first commit to version control,
+we remove `R/hello.R` and `man/hello.Rd`
+(which the project wizard "helpfully" provides as starting points),
+and add `README.md`, `LICENSE.md`, `CONDUCT.md`, and `CITATION.md`
+to describe the project as a whole,
+its license,
+the contributor code of conduct,
+and how we want the project cited.
+These files are nothing to do with R per se
+(and as we'll see below, R isn't entirely happy having them here with these names),
+but as the [previous book discussed](#FIXME),
+every project should have these files.
+
+We then create a placeholder version of the function we want to write:
+
+```{r, child="rse-r-package/etc/frequency-func.R"}
+```
+
+and then make the following changes to `DESCRIPTION`:
+
+1.  change the `Title`, `Author`, `Maintainer`, and `Description`;
+2.  change the `License` to [`MIT`][mit-license] (see FIXME for other options); and
```

Comment:
This should be done automatically with `usethis::use_mit_license()`.

Reply:
done